graph/nullable.md

```diff
@@ -0,0 +1,312 @@
+Below are some examples of nullable and non-nullable properties. 
+
+## CSDL
+
+```xml
+<EntitySet Name="servicePrincipals" Type="self.servicePrincipal" />
+...
+<EntityType Name="servicePrincipal">
+  <Key>
+    <PropertyRef Name="id" />
+  </Key>
+  <Property Name="id" Type="Edm.String" Nullable="false" />
+
+  <Property Name="appId" Type="Edm.String" Nullable="false" /> <!--required for creation-->
+  <Property Name="displayName" Type="Edm.String" Nullable="false" />
+  <Property Name="foo" Type="Edm.String" Nullable="true" DefaultValue="testval" />
+  <Property Name="bar" Type="Edm.String" Nullable="false" DefaultValue="differentvalue" />
+  ...
+</EntityType>
+```
+
+## HTTP Requests
+
+### {1} Create a servicePrincipal with no properties
+
+```HTTP
+POST /servicePrincipals
+
+400 Bad Request
+{
+  "error": {
+    "code": "badRequest",
+    "message": "The 'appId' property is required to create a servicePrincipal."
+  }
+}
+```
+
+### {2} Create a servicePrincipal without a display name
+
+```HTTP
+POST /servicePrincipals
+{
+  "appId": "00000000-0000-0000-0000-000000000001"
+}
+
+201 Created
+{
+  "appId": "00000000-0000-0000-0000-000000000001",
+  "displayName": "some application name",
+  "foo": "testval",
+  "bar": "differentvalue",
+  ...
+}
+```
+Notes:
+1. `displayName` was given a value by the service even though no value was provided by the client
+2. `foo` has the default value as specified by its `DefaultValue` attribute in the CSDL
+3. `bar` has the default value as specified by its `DefaultValue` attribute in the CSDL
+
+### {3} Update the display name of a service principal to null
+
+```HTTP
+PATCH /servicePrincipals/00000000-0000-0000-0000-000000000001
+{
+  "displayName": null
+}
+
+400 Bad Request
+{
+  "error": {
+    "code": "badRequest",
+    "message": "null is not a valid value for the property 'displayName'; 'displayName' is not a nullable property."
+  }
+}
+```
+Notes:
+1. `displayName` cannot be set to `null` because it has be marked with `Nullable="false"` in the CSDL.
+
+### {4} Update the display name of a service principal
+
+```HTTP
+PATCH /servicePrincipals/00000000-0000-0000-0000-000000000001
+{
+  "displayName": "a non-generated display name"
+}
+
+200 OK
+{
+  "appId": "00000000-0000-0000-0000-000000000001",
+  "displayName": "a non-generated display name",
+  "foo": "testval",
+  "bar": "differentvalue",
+  ...
+}
+```
+Notes:
+1. `displayName` can be set to any value other than `null`
+2. The response body here is provided for clarity, and is not part of the guidance itself. The [Graph standard](https://docs.oasis-open.org/odata/odata/v4.01/odata-v4.01-part1-protocol.html#sec_UpdateanEntity) states that the workload can decide the behavior.
+
+### {5} Update the foo property of a service principal to null
```

Comment:
Can we combine some of these examples to make updates to multiple properties?  That would shorten the document.

Reply:
There are 3 dimensions to the properties:
1. Nullable
2. Default value
3. Required for creation

I did at least trim this down from the possible 8 to only 4. 

There are 3 cases per property for each request:
1. Not provided
2. Provided as null
3. Provided as not null

So, I wrote 2 samples for the creation, and then I did cover 100% of the cases per property. 

I see your point about folks filling in the blanks for themselves, but I'm not really sure who would read this particular document top to bottom in the first place. To me, this doc seems more for the exact purpose of covering 100% of cases so that if there's any confusion, we can immediately clear it up.

I'm pushing back on this feedback in particular. If there's something you would like to use this document for where brevity would be better, I can add some notes at the top or something like that to suit those needs.

Reply:
I'm not going to block on this.  This topic is very useful and closes a major gap in API producers understanding in this space, so I'm in favour of publishing and potentially refining and making more concise later.